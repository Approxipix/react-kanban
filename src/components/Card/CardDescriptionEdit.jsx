import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editCardDesc } from '../../redux/cardReducer/actions';
import styled from 'styled-components'
import ClickOutside from '../ClickOutside';

const Form = styled.form`
  width: 90%;
`;

const Input = styled.input`
  width: 100%;
  padding: .3rem .5rem;
  font-size: 1.5rem;
  color: #40424b;
  font-weight: 600;
  border: none;
  box-shadow: none;
`;

class CardDescriptionEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.cardDescription || '',
    };
  }

  handleChange = (key, value) => {
    this.setState({
      ...this.state,
      [key]: value,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { description } = this.state;
    const { cardId } = this.props;
    this.props.actions.editCardDesc({
      cardDescription: description,
      cardId: cardId,
    });
    this.props.toggleIsDescEditing()
  };

  render() {
    const { description } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleIsDescEditing}>
        <Form onSubmit={this.handleSubmit} id="card-edit-title-form">
          <Input
            autoFocus
            type="text"
            placeholder="Edit card description"
            value={description}
            onChange={(e) => this.handleChange('description', e.target.value)}
            spellCheck={false}
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCardDesc: editCardDesc,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CardDescriptionEdit);
