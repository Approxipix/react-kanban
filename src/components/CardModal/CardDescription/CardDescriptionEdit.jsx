import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editCardDesc } from '../../../redux/cardReducer/actions';
import { Actions, CancelButton, SubmitButton, TextArea } from '../../BaseComponent';
import ClickOutside from '../../ClickOutside';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const textareaLineHeight = 16;

class CardDescriptionEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.cardDescription || '',
      rows: 2,
      minRows: 2,
    };
  }

  componentDidMount() {
    this.setState({
      rows: ~~(this.textarea.scrollHeight / textareaLineHeight)
    })
  }

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditDescription()
    }
  };

  handleChange = (e) => {
    const { minRows } = this.state;
    const previousRows = e.target.rows;
    e.target.rows = minRows; // reset number of rows in textarea
    e.target.style.overflow = 'hidden'; //overflow because scrollbar affects e.target.scrollHeight
    const currentRows = ~~(e.target.scrollHeight / textareaLineHeight);
    if (currentRows === previousRows) {
      e.target.rows = currentRows;
    }
    this.setState({
      description: e.target.value,
      rows: currentRows,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { description } = this.state;
    const { cardId } = this.props;
    this.props.actions.editCardDesc({
      cardDescription: description,
      cardId: cardId,
    });
    this.props.toggleEditDescription()
  };

  render() {
    const { description, rows } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditDescription}>
        <form>
          <TextArea
            type="text"
            size=".875rem;"
            padding=".5rem"
            value={description}
            rows={rows}
            placeholder="Edit card description"
            ref={(e) => { this.textarea = e }}
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e)}
            onBlur={() => this.props.toggleEditDescription()}
            autoFocus
            spellCheck={false}
          />
          <Actions>
            <SubmitButton onMouseDown={(e) => this.handleSubmit(e)} >
              Save
            </SubmitButton>
            <CancelButton onClick={() => this.props.toggleEditDescription()}>
              <FontAwesomeIcon icon="times" />
            </CancelButton>
          </Actions>
        </form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCardDesc: editCardDesc,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CardDescriptionEdit);
