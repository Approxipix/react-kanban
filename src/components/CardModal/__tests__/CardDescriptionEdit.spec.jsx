import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardDescriptionEdit from '../CardDescription/CardDescriptionEdit';
import configureStore from "redux-mock-store";
import { number, string } from "prop-types";

const mockStore = configureStore([]);

describe('<CardDescriptionEdit>', () => {
  let store;
  let props;
  let component;
  let mockEditCardDescription = jest.fn();
  let mockToggleEditDescription = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();

    store = mockStore({});
    store.dispatch = mockEditCardDescription;
    props = {
      cardId: 'CardID',
      cardDescription: 'Card Description',
      toggleEditDescription: mockToggleEditDescription
    };

    const wrapper = shallow(<CardDescriptionEdit store={store} {...props} />, { disableLifecycleMethods: true });
    component = wrapper.find('CardDescriptionEdit').dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should call toggleEditDescription if key code equal 27", () => {
    component.find('TextArea').simulate('keydown', { keyCode: 27 });

    expect(mockToggleEditDescription).toHaveBeenCalledTimes(1);
  });

  it('should call handleChange on description change with the correct params', () => {
    const spy = jest.spyOn(component.instance(), "handleChange");
    const mockEvent = {
      target: {
        name: "description",
        value: "Test Description",
        style: {}
      }
    };

    component.find("TextArea").simulate("change", mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(mockEvent);
  });

  it('should call setState on description', () => {
    const mockEvent = {
      target: {
        name: "description",
        value: "Test Description",
        style: {}
      }
    };
    const expectedState = {
      description: "Test Description",
      rows: expect.any(Number),
      minRows: expect.any(Number),
    };

    component.instance().handleChange(mockEvent);

    expect(component.state()).toEqual(expectedState)
  });

  it("should call handleSubmit with the correct params", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.setState({ description: "Test Description" });
    component.instance().handleSubmit(mockEvent);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("should return handleSubmit if description is empty", () => {
    const spy = jest.spyOn(component.instance(), "handleSubmit");
    const mockEvent = {
      preventDefault: jest.fn()
    };

    component.instance().handleSubmit(mockEvent);

    expect(spy).toReturn();
  });

  it("should dispatch an action after handleSubmit", () => {
    const expectedAction = {
      type: 'EDIT_CARD_DESCRIPTION',
      payload: {
        cardDescription: 'Test Description',
        cardId: 'CardID',
      }
    };

    component.setState({ description: "Test Description" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockEditCardDescription).toHaveBeenCalledTimes(1);
    expect(mockEditCardDescription).toHaveBeenCalledWith(expectedAction);
  });

  it("should call toggleEditDescription after submit", () => {
    component.setState({ description: "Test Description" });
    component.instance().handleSubmit({ preventDefault: jest.fn() });

    expect(mockToggleEditDescription).toHaveBeenCalledTimes(1);
  });
});
