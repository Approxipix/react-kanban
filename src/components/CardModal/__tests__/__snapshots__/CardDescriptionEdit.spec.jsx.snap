// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardDescriptionEdit> should render with given state from Redux store 1`] = `
<ClickOutside
  toggleOpened={[MockFunction]}
>
  <form>
    <TextArea
      autoFocus={true}
      onChange={[Function]}
      onKeyDown={[Function]}
      padding=".5rem"
      placeholder="Edit card description"
      rows={2}
      size=".875rem;"
      spellCheck={false}
      type="text"
      value="Card Description"
    />
    <styled.div>
      <SubmitButton
        onClick={[Function]}
      >
        Save
      </SubmitButton>
      <CancelButton
        onClick={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="times"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </CancelButton>
    </styled.div>
  </form>
</ClickOutside>
`;
